### PR TITLE
chromium: fix widevine

### PR DESCRIPTION
**Pull request by jflanglois**
###### Motivation for this change
This change allows widevine to work in chromium (it was previously broken due to a segfault). Newer versions of chromium do not use the libwidevinecdmadapter.so. Instead, libwidevinecdm.so should be installed in the chromium libExec directory.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/applications/networking/browsers/chromium/browser.nix
Comment (jflanglois) [inline pkgs/applications/networking/browsers/chromium/browser.nix]: Could use some suggestions about whether this is the best way to go about this. I've essentially copied the contents of the widevine plugin here.

**Comment by FRidh** [comment]
cc @bendlas

**Comment by JohnAZoidberg** [comment]
I just built it and it works! (At least Netflix plays)
Thanks!

**Comment by aanderse** [comment]
This is broken in 19.03 too, right?

**Comment by jtojnar** [comment]
Is not there a way to avoid building chromium twice (w/ and w/o widevine)?

**Comment by JohnAZoidberg** [comment]
Yes, seems like `pkgs/applications/networking/browsers/chromium/browser.nix` is the same on unstable and 19.03 and on unstable it's definitely broken.

**Comment by jflanglois** [comment]
> Is not there a way to avoid building chromium twice (w/ and w/o widevine)?

Yeah I was wondering about this... That's part of where my question is coming from above, since in principle this is just adding a file to the chromium path.

**Comment by jflanglois** [comment]
I considered `symlinkJoin` as a possibility, but I wanted to prioritize a working fix first.

**Comment by jflanglois** [comment]
I wrote up a version with `symlinkJoin`, but I'm not sure if it works yet because build times are so long (why would chromium vendor all of their dependencies??). I will build over the coming days. I'll leave it up to maintainers/committers to decide if they want to merge this version first or wait to see how the symlink version goes.

**Comment by JohnAZoidberg** [comment]
Can you share the `symlinkJoin` patch please?
Wouldn't that build fast if you use your channel's chromium version?

**Comment by jflanglois** [comment]
> Can you share the `symlinkJoin` patch please?
> Wouldn't that build fast if you use your channel's chromium version?

No, because you still need to apply a patch to the original source code and enable the `enable_widevine` gnFlags. In any case, here's the patch of what I have so far: [symlinkJoin.txt](https://github.com/NixOS/nixpkgs/files/3146163/symlinkJoin.txt)

 It doesn't work, and my guess is that it's because of how chromium decides on a search path for the shared object (I assume that it still looks in the original chromium-unwrapped directory, rather than the result of the `symlinkJoin`).

**Comment by jflanglois** [comment]
For reviewers/committers: I will not continue investigating the option that avoids a rebuild. If the PR seems fine as-is I don't think we should wait for anything else.

**Comment by JohnAZoidberg** [comment]
Yeah, doesn't look like it's possible without rebuild :/
I think this is a big dealbreaker for many people wanting to use chromium. I'd even suggest to have Hydra build it.

What do the maintainers think? Is this ok? @bendlas @ivan

File: pkgs/applications/networking/browsers/chromium/browser.nix
Comment (ivan) [inline pkgs/applications/networking/browsers/chromium/browser.nix]: Should this also have something to work on `beta`, or am I missing something?

**Comment by ivan** [comment]
@JohnAZoidberg I don't believe Hydra could distribute Chrome's widevine module: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=921738#10

**Comment by jflanglois** [comment]
There is one more option I'd like to try to avoid a rebuild which is to copy the contents of the package (rather than create symlinks) and then re-run the necessary patchelf. I think that would work but I haven't had time to try it.

**Comment by jflanglois** [comment]
@bendlas @ivan @JohnAZoidberg I've added a commit that avoids a full rebuild. It seems to work without running patchelf but I'm not very confident with this approach so could use a review.

**Review by ivan — CHANGES_REQUESTED** [review]
Because redistributing `libwidevinecdm.so` is not allowed, I think we need `meta.license` to get set to `licenses.unfree` when `enableWideVine` is set to `true`. `enableWideVine` is false by default, but someone could set it to true and the widevine blob would end up in Hydra, which would be bad.

**Comment by jflanglois** [comment]
@ivan but in this method, we are not redistributing the shared object, right?

**Comment by jflanglois** [comment]
I see. I will make that change. So does this mean that the old method of widevine also had this issue?

**Comment by ivan** [comment]
Yes, I believe so.

File: pkgs/applications/networking/browsers/chromium/default.nix
Comment (ivan) [inline pkgs/applications/networking/browsers/chromium/default.nix]: Let's have the `"beta"` channel written out explicitly here, and throw an error for an unexpected channel.

File: pkgs/applications/networking/browsers/chromium/default.nix
Comment (ivan) [inline pkgs/applications/networking/browsers/chromium/default.nix]: I am confused about this: why did this end up here separately instead of in `browser.nix`'s installPhase? If this must be here, please comment why.

Comment (jflanglois) [reply pkgs/applications/networking/browsers/chromium/default.nix]: This is probably an artifact of when I was playing around with things. I'll try moving it back into `browser.nix`.

Comment (jflanglois) [reply pkgs/applications/networking/browsers/chromium/default.nix]: Ah now I remember, it's because `default.nix` introspects the derivation from `browser.nix` for `passthru.sandboxExecutableName`. I didn't want to try to re-expose everything, but perhaps you have a suggestion for how to deal with that.

Comment (jflanglois) [reply pkgs/applications/networking/browsers/chromium/default.nix]: @ivan additionally, the issue with putting stuff in `installPhase` is that it removes the benefit of minimal rebuild when enabling WV.

Comment (ivan) [reply pkgs/applications/networking/browsers/chromium/default.nix]: @jflanglois that makes sense, can you add a comment to that effect so that someone doesn't refactor it later?

Comment (jflanglois) [reply pkgs/applications/networking/browsers/chromium/default.nix]: @ivan done!

**Review by ivan — APPROVED** [review]
Tested on NixOS x86_64 with the default `enableWideVine` set to `false`,
- confirmed that chromium stable doesn't complain in the stdout or infobars
- confirmed that https://bitmovin.com/demos/drm does not detect widevine.

Tested on NixOS x86_64 with `enableWideVine` set to `true`,
- confirmed no rebuild of all of chromium
- confirmed that chromium stable doesn't complain in the stdout or infobars
- confirmed that https://bitmovin.com/demos/drm detects widevine
- confirmed that license is set to unfree

I have also been using earlier versions of this PR and confirmed that Netflix worked fine as well.

**Comment by jflanglois** [comment]
Thanks @ivan for the review! One other thing that I tested as well is to install chromium with widevine and than run garbage collection. Chromium continued to work after GC. I did this out of paranoia that necessary dependencies wouldn't be held onto with the `runCommand` derivation.

**Comment by ivan** [comment]
I also tested `chromiumBeta` and `chromiumDev` now, both with and without `enableWideVine`, and everything seems to work as expected.

@bendlas would you like to review or merge?

**Comment by bendlas** [comment]
Thanks!

**Comment by bendlas** [comment]
I backported this to 19.09 https://github.com/NixOS/nixpkgs/commit/d1cd6f8db3eaf6f282989820252ded3091638174

The approach looks good to me and this has been a long-standing problem. I think it's worth pushing out for this release.

**Comment by jflanglois** [comment]
Thank you!

**Pull request merged**
